c/extras/aggregator.cc

```diff
@@ -427,35 +432,49 @@ void Aggregator::group_nd(DataTablePtr& dt_exemplars, DataTablePtr& dt_members)
 
 
 /*
-*  Adjust `delta` (i.e. `radius^2`) based on the closest exemplar to the first one
-*  and merge all the exemplars within that distance.
+*  Adjust `delta` (i.e. `radius^2`) based on the mean distance between
+*  the gathered exemplars and merge all the exemplars within that distance.
 */
 void Aggregator::adjust_delta(double& delta, std::vector<ExPtr>& exemplars,
                               std::vector<int64_t>& ids, int64_t ndims) {
-  double min_distance = std::numeric_limits<double>::max();
-
-  for (auto it1 = exemplars.begin(); it1 < exemplars.end() - 1; ++it1) {
-    std::vector<ExPtr>::iterator min_it = exemplars.end();
-    for (auto it2 = it1 + 1; it2 < exemplars.end(); ++it2) {
-      double distance = calculate_distance((*it1)->coords,
-                                           (*it2)->coords,
-                                           ndims,
-                                           delta,
-                                           it1 != exemplars.begin());
-      if (distance < min_distance ) {
-        if (it1 == exemplars.begin()){
-          min_distance = distance;
-          delta += min_distance;
-        }
-        min_it = it2;
-      }
+  using namespace std;
+
+  size_t n = exemplars.size();
+  size_t n_distances = (n * n - n) / 2;
+  double total_distance = 0.0;
+
+  map<pair<size_t, size_t>, double> deltas;
+
+  for (size_t i = 0; i < n - 1; ++i) {
+    for (size_t j = i + 1; j < n; ++j) {
+      double d = calculate_distance(exemplars[i]->coords,
+                                     exemplars[j]->coords,
+                                     ndims,
+                                     delta,
+                                     0);
+      deltas.insert(make_pair(make_pair(exemplars[i]->id,exemplars[j]->id), d));
+      total_distance += sqrt(d);
     }
+  }
 
-    if (min_it != exemplars.end()) {
-      ids[static_cast<size_t>((*min_it)->id)] = (*it1)->id;
-      exemplars.erase(min_it);
+  double delta_new = pow(0.5 * total_distance / n_distances, 2);
+
+  auto it1 = exemplars.begin();
```

Comment:
Thanks, yes, I know this could be refactored. For the moment number of exemplars is very small ~250, so even with this approach it takes almost no time comparing to the aggregation time.

Reply:
Btw, with your approach one will save on moving the data, however, I'm not sure it will help avoid using maps. Because every time one removed an exemplar (sets it to be `nullptr`) "a simple formula mapping triangular index (i, j) into an index within this array" will have to be updated in a non-trivial way.

Reply:
Also, with this approach we will have to introduce quite many `if`s to check weather or not a particular exemplar is `nullptr`:
1) when a row is tested against a set of exemplars and the distance is calculated ~ `nrows` times;
2) when distances between all the exemplars are calculated and exemplars are "removed" ~ `n_exemplars * n_exemplars`.

Reply:
Formula: `idx(i,j) = (2*n-i-1)*i/2 + j`. As long as indices of exemplars do not change (which is achieved by having tombstones), the mapping from (i,j) into this idx remains fixed.
We don't do row testing within this function, so why is (1) relevant? And at the end of this function, the array of exemplars is dense again.

Reply:
The formula assumes `i` runs from `0` to `n-2`, and `j` from `i` to `n-1`.

Reply:
Ah, so your point was to `compatify` it at the end of this particular function? Then this is ok, of course. I was like you propose do this once, only at the end of the aggregation. 

Actually it could be re-factored even further, just walking through the array of distances and calculate `(i, j)` from the array index `k`.